src/main.js

```diff
@@ -0,0 +1,187 @@
+      schemaPrefix:  'cloud-mirror/v1/',
+      aws:           cfg.aws
+    })
+  },
+
+  /*publisher: {
+    requires: ['cfg', 'validator', 'influx', 'process'],
+    setup: ({cfg, validator, influx, process}) => exchanges.setup({
+      credentials:        cfg.pulse,
+      exchangePrefix:     cfg.app.exchangePrefix,
+      validator:          validator,
+      referencePrefix:    'cloud-mirror/v1/exchanges.json',
+      publish:            cfg.app.publishMetaData,
+      aws:                cfg.aws,
+      drain:              influx,
+      component:          cfg.app.statsComponent,
+      process:            process
+    })
+  },*/
+
+  api: {
+    requires: [
+      'cfg', /*'publisher',*/ 'validator', 'influx', 'memcached', 's3backends',
+    ],
+    setup: (ctx) => v1.setup({
+      context: {
+        //publisher: ctx.publisher,
+        validator: ctx.validator,
+        memcached: ctx.memcached,
+        s3backends: ctx.s3backends,
+      },
+      validator: ctx.validator,
+      authBaseUrl: ctx.cfg.taskcluster.authBaseUrl,
+      publish: ctx.cfg.app.publishMetaData,
+      baseUrl: ctx.cfg.server.publicUrl + '/v1',
+      referencePrefix: 'cloud-mirror/v1/api.json',
+      aws: ctx.cfg.aws,
+      component: ctx.cfg.app.statsComponent,
+      drain: ctx.influx
+    })
+  },
+
+  // Create the server process
+  server: {
+    requires: ['cfg', 'api', 'monitor'],
+    setup: ({cfg, api}) => {
+      let app = base.app(cfg.server);
+      app.use('/v1', api);
+      return app.createServer();
+    }
+  },
+
+  s3backends: {
+    requires: ['cfg', 'sqs', 'memcached', 'profile'], 
+    setup: async ({cfg, sqs, memcached, profile}) => {
+      // This should probably not all be here...
+      let s3objs = {};
+      let s3buckets = {};
+      let s3backends = {};
+      let regions = cfg.backend.s3.regions.split(',');
+
+      // First, let's asynchronusly create the buckets we need
+      await Promise.all(regions.map(region => {
+        let awsCfg = _.omit(cfg.aws, 'region');
+        awsCfg.region = region;
+        let s3 = new aws.S3(awsCfg);
+        s3objs[region] = s3;
+        let bucket = cfg.backend.s3.bucketBase + cfg.server.env;
+        bucket += '-' + region;
+        s3buckets[region] = bucket;
+        return s3Backend.createS3Bucket(s3, bucket, region, 
+                                       cfg.backend.s3.acl,
+                                       cfg.backend.s3.lifespan);
+      }));
+
+      // Now, let's instantiate the backends we need!
+      for (let region of regions) {
+        let backend = new s3Backend.S3Backend({
+          region: region,
+          bucket: s3buckets[region],
+          memcachedTTL: cfg.backend.memcachedTTL,
+          sqs: sqs,
+          s3: s3objs[region],
+          memcached: memcached,
+          redirectLimit: cfg.app.redirectLimit,
+          ensureSSL: cfg.app.ensureSSL,
+          allowedPatterns: cfg.app.allowedPatterns.map(x => new RegExp(x)),
```

Comment:
Can we agree `allowedPatterns` does not belong here? it's it better just to do this in the api definition?

I thinking it would reduce code in backend, making it less complicated. And we could send better error messages.